### PR TITLE
[C10D] Implement Store fallbacks for append, multi_get and multi_set.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100768
* #100742

These fallbacks exposed some issue in quite a few spots in our bindings.

### CODE REVIEW

File: torch/csrc/distributed/c10d/Store.cpp [hunk 1]
```diff
@@ -38,18 +38,34 @@ std::string Store::get_to_str(const std::string& key) {
...
+  // This fallback depends on compareSet
+  std::vector<uint8_t> expected = value;
+  std::vector<uint8_t> current;
+  // cannot use get(key) as it might block forever if the key doesn't exist
+  current = compareSet(key, current, expected);
+  while (current != expected) {
+    expected = current;
+    expected.insert(expected.end(), value.begin(), value.end());
+    current = compareSet(key, current, expected);
+  }
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/Store.cpp:50]: How can we prevent forever loop?

File: torch/csrc/distributed/c10d/Store.cpp [hunk 2]
```diff
@@ -38,18 +38,34 @@ std::string Store::get_to_str(const std::string& key) {
 }
 
 void Store::append(const std::string& key, const std::vector<uint8_t>& value) {
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/Store.cpp:40]: This is a fallback when `has_extended_api` is false right?

**Review by fduwjj — APPROVED** [review]
Unblock.

File: torch/csrc/distributed/c10d/Store.cpp (hunk 1 above)
Comment (kumpera) [reply torch/csrc/distributed/c10d/Store.cpp:50]: In theory, we can't prevent infinite loops when using CAS as long as there are multiple active clients going at each other.

OTOH, we don't have spurious, no progress, CAS failures so each time it fails there has being global progress.

File: torch/csrc/distributed/c10d/Store.cpp (hunk 2 above)
Comment (kumpera) [reply torch/csrc/distributed/c10d/Store.cpp:40]: Yup, these fallbacks are in place when the underlying store doesn't implement the newer methods.

**Pull request closed without merging**